src/Microsoft.Diagnostics.Monitoring/Pipeline.cs

```diff
@@ -0,0 +1,220 @@
+                }
+            }
+            return runTask;
+        }
+
+        private async Task RunAsyncCore(CancellationToken token)
+        {
+            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _disposeSource.Token))
+            {
+                try
+                {
+                    linkedSource.Token.ThrowIfCancellationRequested();
+                    await OnRun(linkedSource.Token);
+                }
+                finally
+                {
+                    await Cleanup();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Causes an unstarted or running pipeline to transition to the stopping state. In this
+        /// state data flow from the source will be stopped and any in-flight data is gracefully
+        /// drained. Calling this more than once doesn't have any additional effect and returns
+        /// the same Task. Once the pipeline transitions to the Stopped state the returned Task
+        /// will be complete or cancelled.
+        /// </summary>
+        /// <param name="cancelToken">If this token is cancelled, it aborts the operation without consideration for
+        /// preserving the data</param>
+        /// <exception cref="PipelineException">For any error that prevents all the requested
+        /// data from being moved through the pipeline</exception>
+        /// <remarks>Any exception other than PipelineException represents either
+        /// a bug in the pipeline implementation because it was unanticipated or a failure in
+        /// lower level runtime/OS/hardware to keep the process in a consistent state</remarks>
+        public Task StopAsync(CancellationToken token = default)
+        {
+            Task stopTask = null;
+            lock (_lock)
+            {
+                if (_isCleanedUp)
+                {
+                    stopTask = _stopTask ?? Task.CompletedTask;
+                }
+                else if (_runTask == null)
+                {
+                    throw new PipelineException("Unable to stop unstarted pipeline");
+                }
+                else
+                {
+                    if (_stopTask == null)
+                    {
+                        _stopTask = StopAsyncCore(token);
+                    }
+                    stopTask = _stopTask;
+                }
+            }
+            return stopTask;
+        }
+
+        private async Task StopAsyncCore(CancellationToken token)
+        {
+            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _disposeSource.Token))
+            {
+                try
+                {
+                    linkedSource.Token.ThrowIfCancellationRequested();
+                    await OnStop(linkedSource.Token);
+                }
+                catch (OperationCanceledException)
+                {
+                    await Cleanup();
+                    throw;
```

Comment:
nit: might want to call `token.ThrowIfCancellationRequested()` here before calling `throw` because the caller might want to check if the unwinding OperationCanceledException was caused by them cancelling their token or by something else e.g. `catch (OperationCanceledException ex) when (ex.CancellationToken == token)`.